megaqc/rest_api/plot.py

```diff
@@ -27,129 +29,90 @@ def trend_data(fields, filter, plot_prefix, control_limits, center_line):
-            .with_entities(
-                models.Sample.sample_name,
-                models.SampleDataType.nice_name,
-                models.Report.created_at,
-                models.SampleData.value,
-            )
-            .order_by(
-                models.Report.created_at.asc(),
-            )
-            .filter(Sample.sample_id.in_(subquery))
-            .distinct()
+    # Choose the columns to select, and further filter it down to samples with the column we want to plot
+    query = (
+        db.session.query(Sample)
+        .join(SampleData, isouter=True)
+        .join(SampleDataType, isouter=True)
+        .join(Report, Report.report_id == Sample.report_id, isouter=True)
+        .with_entities(
+            models.Sample.sample_name,
+            models.SampleDataType.nice_name,
+            models.Report.created_at,
+            models.SampleData.value,
         )
+        .order_by(
+            models.Report.created_at.asc(),
+        )
+        .filter(Sample.sample_id.in_(subquery))
+        .distinct()
+    )
 
-        # Fields can be specified either as type IDs, or as type names
-        if field.isdigit():
-            query = query.filter(models.SampleDataType.sample_data_type_id == field)
-        else:
-            query = query.filter(models.SampleDataType.data_key == field)
-
-        data = query.all()
-
-        # If the query returned nothing, skip this field
-        if len(data) == 0:
-            break
-
-        names, data_types, x, y = zip(*data)
-        data_type = data_types[0]
-        names = numpy.asarray(names, dtype=str)
-        x = numpy.asarray(x)
-        y = numpy.asarray(y, dtype=float)
-
-        # Anything outside the control limits is an outlier
-        outliers = absolute(zscore(y)) > control_limits["sigma"]
-        inliers = ~outliers
-
-        # Add the outliers
-        plots.append(
-            dict(
-                id=plot_prefix + "_outlier_" + field,
-                type="scatter",
-                text=names[outliers],
-                hoverinfo="text+x+y",
-                x=x[outliers],
-                y=y[outliers],
-                line=dict(color="rgb(250,0,0)"),
-                mode="markers",
-                name="{} Outliers".format(data_type),
-            )
+    # Fields can be specified either as type IDs, or as type names
+    if fields[0].isdigit():
+        query = query.filter(models.SampleDataType.sample_data_type_id.in_(fields))
+    else:
+        query = query.filter(models.SampleDataType.data_key.in_(fields))
+
+    data = query.all()
+
+    names, data_types, x, y = zip(*data)
+    data_type = data_types[0]
+    names = numpy.asarray(names, dtype=str)
+    x = numpy.asarray(x)
+    y = numpy.asarray(y, dtype=float)
+
+    cov = EmpiricalCovariance()
+    y = y.reshape(-1, len(fields))
+
+    cov.fit(y)
+    distance = cov.mahalanobis(y)
+
+    # Calculate the critical value
+    n, p = y.shape
+    cri = f.isf(0.05, dfn=p, dfd=n - p)
```

Comment:
Add a comment here explaining what you are doing.

Reply:
Calculate the line of critical value for the distribution.

Reply:
Yes, but some more detail would be good explain the statistics being used here. You might want to refer to: https://online.stat.psu.edu/stat505/lesson/7/7.1/7.1.3

Reply:
Using Hotelling’s T-Square to test the statistic for the F distribution.